controllers/usernamespace/controller_test.go

```diff
@@ -0,0 +1,361 @@
+		}
+		if len(cs.Items) > 0 {
+			t.Errorf("No configmaps expected in the tested namespace but found %d", len(cs.Items))
+		}
+	}
+
+	t.Run("k8s", func(t *testing.T) {
+		test(t, infrastructure.Kubernetes, &corev1.Namespace{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: "ns",
+			},
+		})
+	})
+
+	t.Run("openshift", func(t *testing.T) {
+		test(t, infrastructure.OpenShiftv4, &projectv1.Project{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: "prj",
+			},
+		})
+	})
+}
+
+func TestRequiresLabelsToMatchOneOfMultipleCheCluster(t *testing.T) {
+	test := func(t *testing.T, infraType infrastructure.Type, namespace metav1.Object) {
+		ctx := context.TODO()
+		scheme, cl, r := setup(infraType, namespace.(runtime.Object))
+		setupCheCluster(t, ctx, cl, scheme, "che1", "che")
+		setupCheCluster(t, ctx, cl, scheme, "che2", "che")
+
+		res, err := r.Reconcile(reconcile.Request{NamespacedName: types.NamespacedName{Name: namespace.GetName()}})
+		if err != nil {
+			t.Fatal(err)
+		}
+
+		if !res.Requeue {
+			t.Error("The reconciliation request should have been requeued.")
+		}
+	}
+
+	t.Run("k8s", func(t *testing.T) {
+		test(t, infrastructure.Kubernetes, &corev1.Namespace{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: "ns",
+				Labels: map[string]string{
+					workspaceNamespaceOwnerUidLabel: "uid",
+				},
+			},
+		})
+	})
+
+	t.Run("openshift", func(t *testing.T) {
+		test(t, infrastructure.OpenShiftv4, &projectv1.Project{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: "prj",
+				Labels: map[string]string{
+					workspaceNamespaceOwnerUidLabel: "uid",
+				},
+			},
+		})
+	})
+}
+
+func TestMatchingCheClusterCanBeSelectedUsingLabels(t *testing.T) {
+	test := func(t *testing.T, infraType infrastructure.Type, namespace metav1.Object) {
+		ctx := context.TODO()
+		scheme, cl, r := setup(infraType, namespace.(runtime.Object))
+		setupCheCluster(t, ctx, cl, scheme, "che1", "che")
+		setupCheCluster(t, ctx, cl, scheme, "che2", "che")
+
+		res, err := r.Reconcile(reconcile.Request{NamespacedName: types.NamespacedName{Name: namespace.GetName()}})
+		if err != nil {
+			t.Fatal(err)
+		}
+
+		if res.Requeue {
+			t.Error("The reconciliation request should have succeeded but is requesting a requeue.")
+		}
+	}
+
+	t.Run("k8s", func(t *testing.T) {
+		test(t, infrastructure.Kubernetes, &corev1.Namespace{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: "ns",
+				Labels: map[string]string{
+					workspaceNamespaceOwnerUidLabel: "uid",
+					cheNameLabel:                    "che",
+					cheNamespaceLabel:               "che1",
+				},
+			},
+		})
+	})
+
+	t.Run("openshift", func(t *testing.T) {
+		test(t, infrastructure.OpenShiftv4, &projectv1.Project{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: "prj",
+				Labels: map[string]string{
+					workspaceNamespaceOwnerUidLabel: "uid",
+					cheNameLabel:                    "che",
+					cheNamespaceLabel:               "che1",
+				},
+			},
+		})
+	})
+}
+
+func TestCreatesDataInNamespace(t *testing.T) {
```

Comment:
I believe tests readability would benefit if you separate it into three tests: TestPropagatingTrutedCerts, TestPropagatingSelfSignedCert, TestPropagatingProxySettings.

Reply:
True, but let's wait with that until we have even more data synced....